benchmark/C/Savina/src/concurrency/Dictionary.lf

```diff
@@ -0,0 +1,203 @@
+};
+
+preamble {=
+    #include <map>
+    #include <vector>
+    #include <memory>
+    enum class AccessType {
+      Read,
+      Write
+    };
+    
+    struct Message {
+      AccessType type;
+      int key;
+      int value;
+    };
+=}
+
+reactor Manager(numWorkers: size_t(20)) {
+    
+    state numWorkersTerminated: size_t(0);
+    
+    input start: bool;
+    output finished: bool;
+    
+    output doWork: bool;
+    input[numWorkers] workerFinished: bool;
+    
+    reaction(startup) -> doWork {=
+        //In C: initialize local state
+        self->numWorkersTerminated = 0;
+        // start execution
+        SET(doWork, true);
+    =}
+    
+    reaction(workerFinished) -> finished {=
+        for (int i = 0;  i < self->numWorkers; i ++) {
+            if (workerFinished[i]->is_present) {
+                self->numWorkersTerminated++;
+            }
+        }
+        if(self->numWorkersTerminated == self->numWorkers) {
+            printf("All workers terminated.");
+            SET(finished, true);
+        }
+    =}
+}
+
+reactor DictionaryImpl(numWorkers: size_t(20)) {
+    
+    
+    state dataMap: std::map<int,int>;
+    state answersToSend: std::vector<int>;
+    
+    // Having the action in the dictionary is fatser...
+    logical action sendAnswers;
+
+    input[numWorkers] request: Message*;
+    output[numWorkers] response: int; 
+    
+    reaction(startup) {=
+        // in C: initialize local state
+        self->answersToSend.resize(self->numWorkers, -1);
+    =}
+    
+    reaction(sendAnswers) -> response {=
+        for(size_t i = 0; i < self->numWorkers; i++) {
+            if(self->answersToSend[i] >= 0) {
+                SET(response[i], self->answersToSend[i]);
+                self->answersToSend[i] = -1;
+            }
+        }
+    =}
+    
+    reaction(request) -> sendAnswers {=
+        // The order of messages to read is relevant, it effectively
+        // assigns priorities to the workers.
+        for(size_t i = 0; i < self->numWorkers; i++) {
+            if(request[i]->is_present) {
+                auto msg = request[i]->value;
+                
+                if(msg->type == AccessType::Write) {  
+                    self->dataMap.emplace(msg->key, msg->value);
+                    // Savina sends ResultMsg always independently if adding (key,value)
+                    // to the map was successful.
+                    self->answersToSend[i] = msg->value;
+                } else if(msg->type == AccessType::Read) {
+                    // read the value. If the key is not yet present, this will add it and initialize the value to 0.
+                    // This does not match 100% the Savina implementation which returns null if the key is not present,
+                    // but should do a similar job.
+                    int value = self->dataMap[msg->key];
+                    self->answersToSend[i] = value;
+                }
+                //FIXME: free the message here?
```

Comment:
I think the C runtime interprets the type `Message *` as a token and implements its own mechanism to free it. Could @Soroosh129 or @edwardalee help here? The program currently segfaults. I am not sure if this is the issue, though.

Reply:
I looked into this and still can't figure out what would be causing the segfault.  I tried putting the Messages on the stack and allocating memory for some of the local reactor states but that doesn't work (plus, putting Messages on the stack will likely cause memory problems if we ever want to test the performance on larger problem sizes).  I also tried letting the C runtime free the Messages, per your comment, but that also segfaults.  The only thing I can think of now would be the std::map<int, int> used in the Dictionary reactor causing this, but that map is one of the major reasons we are using the CCpp target- so, I'm not sure what the most elegant way around this is...

Reply:
Indeed the problem was with the map, but I don't think it is a problem with the map itself, but rather with how it is initialized (or not) in the CCpp target. See #741.